llvm/lib/Transforms/Yk/ControlPoint.cpp

```diff
@@ -88,36 +89,149 @@ CallInst *findControlPointCall(Module &M) {
+    } else {
+      // Terminating instruction: successor instructions are the first
+      // instructions of all successor blocks.
+      for (unsigned SuccIdx = 0; SuccIdx < Term->getNumSuccessors(); SuccIdx++)
+        SuccInsts.insert(&*Term->getSuccessor(SuccIdx)->begin());
+    }
+    return SuccInsts;
   }
 
-  // Then add anything which dominates the control point to the live set.
-  for (auto &BB : *Func) {
-    if (!DT.dominates(cast<Instruction>(OldCtrlPoint), &BB)) {
-      for (auto &I : BB) {
-        if ((!I.getType()->isVoidTy()) &&
-            (DT.dominates(&I, cast<Instruction>(OldCtrlPoint)))) {
-          Vec.push_back(&I);
+  /// Replaces the value set behind the pointer `S` with the value set `R` and
+  /// returns whether the set behind `S` changed.
+  bool updateValueSet(std::set<Value *> *S, const std::set<Value *> R) {
+    bool Changed = (*S != R);
+    *S = R;
+    return Changed;
+  }
+
+public:
+  LivenessAnalysis(Function *Func) {
+    // Compute defs and uses for each instruction.
+    std::map<Instruction *, std::set<Value *>> Defs;
+    std::map<Instruction *, std::set<Value *>> Uses;
+    for (BasicBlock &BB : *Func) {
+      for (Instruction &I : BB) {
+        // Record what this instruction defines.
+        if (!I.getType()->isVoidTy())
+          Defs[&I].insert(cast<Value>(&I));
+
+        // Record what this instruction uses.
+        //
+        // Note that Phi nodes are special and must be skipped. If we consider
+        // their operands as uses, then Phi nodes in loops may use variables
+        // before they are defined, and this messes with the algorithm.
+        //
+        // The book doesn't cover this quirk, as it explains liveness for
+        // non-SSA form, and thus doesn't need to worry about Phi nodes.
+        if (isa<PHINode>(I))
+          continue;
+
+        for (auto *U = I.op_begin(); U < I.op_end(); U++) {
+          if ((!isa<Constant>(U)) && (!isa<BasicBlock>(U)) &&
+              (!isa<MetadataAsValue>(U)) && (!isa<InlineAsm>(U))) {
+            Uses[&I].insert(*U);
+          }
         }
       }
     }
+
+    // A function implicitly defines its arguments.
+    //
+    // To propagate the arguments properly we pretend that the first instruction
+    // in the entry block defines the arguments.
+    Instruction *FirstInst = &*Func->getEntryBlock().begin();
+    for (auto &Arg : Func->args())
+      Defs[FirstInst].insert(&Arg);
+
+    // Compute the live sets for each instruction.
+    //
+    // This is the fix-point of the following data-flow equations (page 206 in
+    // the book referenced above):
+    //
+    //    in[I] = use[I] ∪ (out[I] - def[I])
+    //
+    //    out[I] =       ∪
+    //             (S in succ[I])    in[S]
+    //
+    // Note that only the `Ins` map is kept after this constructor ends, so
+    // only `Ins` is a field.
+    std::map<Instruction *, std::set<Value *>> Outs;
+    bool Changed;
+    do {
+      Changed = false;
```

Comment:
Wouldn't the `bool` definition be better here rather than outside the loop?

Reply:
You'd have thought so, but moving it in means that it's out of scope for the enclosing `do... while()`:

```diff
--- a/llvm/lib/Transforms/Yk/ControlPoint.cpp
+++ b/llvm/lib/Transforms/Yk/ControlPoint.cpp
@@ -194,9 +194,8 @@ public:
     // Note that only the `Ins` map is kept after this constructor ends, so
     // only `Ins` is a field.
     std::map<Instruction *, std::set<Value *>> Outs;
-    bool Changed;
     do {
-      Changed = false;
+      bool Changed = false;
       // As the book explains, fixed-points are reached quicker if we process
       // control flow in "approximately reverse direction". Hence these two
       // loops iterate in reverse.
```

```
/home/vext01/research/ykllvm/llvm/lib/Transforms/Yk/ControlPoint.cpp:223:14: error: ‘Changed’ was not declared in this scope
  223 |     } while (Changed); // Until a fixed-point.
      |              ^~~~~~~
```